Scan Delete Support Part 4: Delete File Loading; Skeleton for Processing

sdd opened the pull request:
Extends the `DeleteFileManager` introduced in https://github.com/apache/iceberg-rust/pull/950 To include loading of delete files, storage and retrieval of parsed delete files from shared state, and the outline for how parsing will connect up to this new work.


Builds on top of https://github.com/apache/iceberg-rust/pull/951 and https://github.com/apache/iceberg-rust/pull/950 and should be merged after them.

jonathanc-n commented on crates/iceberg/src/arrow/delete_file_manager.rs:
```diff
@@ -0,0 +1,564 @@
+struct DeleteFileManagerState {
+    // delete vectors and positional deletes get merged when loaded into a single delete vector
+    // per data file
+    delete_vectors: HashMap<String, RoaringTreemap>,
```
I think a function should be included for enabling deletion vectors for the future when a property is added.

sdd replied:
Sorry @jonathanc-n, I'm a bit confused as to what you mean, could you explain?

sdd replied:
I'm resolving this as it's been unanswered for a while and the PR has changed quite significantly since the comment was made. Feel free to add another comment in it's place if you have an update! :-)

jonathanc-n commented on crates/iceberg/src/arrow/delete_file_manager.rs:
```diff
@@ -0,0 +1,564 @@
+        concurrency_limit_data_files: usize,
+    ) -> Result<()> {
+        /*
+               * Create a single stream of all delete file tasks irrespective of type,
```
This is really nice. Are we able to make this into rust doc instead? would be nice for people looking into the crate

sdd replied:
Done!

jonathanc-n commented on crates/iceberg/src/arrow/delete_file_manager.rs:
```diff
@@ -0,0 +1,564 @@
+            )),
+
+            DataContentType::EqualityDeletes => {
+                let (sender, fut) = EqDelFuture::new();
```
I think we can do the check before creating the future

sdd replied:
Done.

jonathanc-n reviewed (commented):
This is nice, will look at the parsed records next.

sdd commented:
@liurenjie1024, @Xuanwo, @Fokko - this is ready for re-review, if you could take a look that would be great!

sdd commented:
Thanks for the review @liurenjie1024 - much appreciated. Will come back with a revised design.

sdd commented on crates/iceberg/src/delete_vector.rs:
```diff
@@ -38,6 +40,10 @@ impl DeleteVector {
         }
     }
+
+    pub fn intersect_assign(&mut self, other: &DeleteVector) {
```
made it `pub(crate)` for now

sdd commented on crates/iceberg/src/arrow/delete_file_manager.rs:
```diff
@@ -47,47 +60,533 @@ impl DeleteFileManager for CachingDeleteFileManager {
+    ) -> Result<Predicate> {
+        // TODO
+
+        Ok(AlwaysTrue)
```
Done

sdd commented on crates/iceberg/src/arrow/delete_file_manager.rs:
```diff
@@ -47,47 +60,533 @@ impl DeleteFileManager for CachingDeleteFileManager {
+    ) -> Result<HashMap<String, DeleteVector>> {
+        // TODO
+
+        Ok(HashMap::default())
```
Done

sdd commented on crates/iceberg/src/arrow/delete_file_manager.rs:
```diff
@@ -47,47 +60,533 @@ impl DeleteFileManager for CachingDeleteFileManager {
+    /// Parses record batch streams from individual equality delete files
+    ///
+    /// Returns an unbound Predicate for each batch stream
+    async fn parse_equality_deletes_record_batch_stream(
```
Seeing as this is around efficiency rather than correctness, would you be ok with deferring a more efficient implementation to a follow-up PR? I'd prefer to have some correct support for delete files sooner and then refine it to perform better afterwards, if that is ok with you.

sdd commented on crates/iceberg/src/arrow/delete_file_manager.rs:
```diff
@@ -47,47 +60,533 @@ impl DeleteFileManager for CachingDeleteFileManager {
+           Essentially a super-cut-down ArrowReader. We can't use ArrowReader directly
+           as that introduces a circular dependency.
+        */
+        let record_batch_stream = ArrowReader::create_parquet_record_batch_stream_builder(
```
Updated the PR to add eq del record batch schema evolution in this commit:

https://github.com/apache/iceberg-rust/pull/982/commits/58b0f075964b3794d50d202a3a474f19947df5f6

sdd commented on crates/iceberg/src/arrow/delete_file_manager.rs:
```diff
@@ -47,47 +60,533 @@ impl DeleteFileManager for CachingDeleteFileManager {
+                predicate.clone()
+            };
+
+            combined_predicate = combined_predicate.and(predicate.await);
```
I've updated to make use of the equality_ids passed through the file scan tasks to ensure that only columns whose field ids are present in a given delete file's eq_ids are included in the predicate: https://github.com/apache/iceberg-rust/pull/1017/commits/f62ac3f98942707eeb7a06410a52a9a5eea951ac#diff-7d31974b51961466c4be7da66f05f93972ac55f35b7f78252aa6df3a6d087ea4R440

sdd commented:
Back to you @liurenjie1024 - I've made the changes around missing functionality. Still the open question of if you are ok to defer the structural / performance changes to a follow-up so that we can make more incremental progress rather than adding yet more changes into this PR.

sdd commented:
Actually @liurenjie1024 I'll go ahead with the structural changes to split this into separate loader and filter structs and update this PR.

sdd commented:
> Thanks @sdd for this pr. There are some missing points in current design. Also I would suggest not putting too much in `DeleteFilterManager`. I suppose `DeleterFilterManager` acting more like a delete loader, which manages the io and caching of record batch. The actual filtering part, could delegate to `DeleteFilter`, WDYT? I think a good reference implementation is java's DeleteFilter, see https://github.com/apache/iceberg/blob/af8e3f5a40f4f36bbe1d868146749e2341471586/data/src/main/java/org/apache/iceberg/data/DeleteFilter.java#L50

I've refactored as you suggested and you're right, it is a better design.

sdd commented:
> Hi, @sdd I saw you opened a series of pr for handling reading of deletions. I have a suggestion about it, instead of opening a series prs, you can have a large draft pr containing all your changes, while pick one component to open a small pr for review. This way reviewer could understand your whole design by walking through the large pr, and review carefully small pr. Also, when reviewer have comments, you only need to change one large pr instead of several small ones, WDYT?

I Could do, if you think it's worth it - the other two remaining PRs after this one are much smaller and it feels like just as much work to merge those into a single PR and then break PRs out of that again as it does to simply keep rebasing those two PRs on top of this one.

I'll be very glad once this delete file read support is done - it's been a long, hard slog to be honest and I'm struggling to stay motivated with it, but we're not far off now, hopefully.

sdd commented:
I had a bug in here that was causing the tests to deadlock in the follow-up PRs. I was missing a waker for my custom futures.That's been rectified now and this PR plus the two follow-ups are now ready for review once more.

sdd commented on crates/iceberg/src/arrow/delete_file_loader.rs:
````diff
@@ -0,0 +1,530 @@
+    ///                                               |
+    ///                                            [join!]
+    /// ```
+    pub(crate) fn load_deletes(
````
It is used. https://github.com/apache/iceberg-rust/pull/982/files#diff-43d2385894ffe1fb08773bf0a7463850036dca5e3252c18d0fc9fe03c79d8a89R201

sdd commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -63,6 +63,7 @@ pub struct ArrowReaderBuilder {
     concurrency_limit_data_files: usize,
     row_group_filtering_enabled: bool,
     row_selection_enabled: bool,
+    delete_file_support_enabled: bool,
```
I'm happy to remove it, I just wanted it in here temporarily until the delete file support was complete with the follow-up PRs, at which point it will be removed. I didn't want to expose users to partially implemented delete file support by default, so I gated it so that is opt-in whilst still in development.

sdd replied:
Done 👍🏼 Delete file processing is always enabled now.

sdd commented on crates/iceberg/src/arrow/delete_file_loader.rs:
```diff
@@ -0,0 +1,530 @@
+pub(crate) struct CachingDeleteFileLoader {
+    file_io: FileIO,
+    concurrency_limit_data_files: usize,
+    del_filter: DeleteFilter,
```
Aah - you're right! This was a holdover from the design before the refactor but it can also be removed if we create the DeleteFilter inside `CachingDeleteFileLoader::load_deletes()`. I've updated the PR to address this.

sdd commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -104,18 +106,25 @@ impl ArrowReaderBuilder {
             batch_size: self.batch_size,
             file_io: self.file_io.clone(),
-            delete_file_manager: CachingDeleteFileManager::new(
+            delete_file_loader: CachingDeleteFileLoader::new(
```
We're not making best use of the features that ArrowRecordBatchReader provides us if we do that. We can't implement the delete filter as a simple filter on a RecordBatchStream unless we ditch using parquet's `ParquetRecordBatchStream`. 

We're using `parquet`'s `ParquetRecordBatchStream` to do the predicate filtering before we even get access to the `RecordBatchStream`. So by the time we have a stream of RecordBatches, we can't apply positional deletes or delete vectors because we no longer know what row number in the original file a record batches row corresponds to, as some rows can have been filtered out.

sdd replied:
I may be able to update ArrowReader to store a DeleteFileFilter rather than a DeleteFileLoader. But not to change the semantics of how the filter itself is used to match `deleteFilter.filter(recordBatchStream)`. Not without rewriting the ArrowReader entirely not use `ParquetRecordBatchStream` and as a consequence needing to reimplement all the predicate filtering logic, row selection, projection, page skipping, and row group skipping that it gives us.

sdd commented on crates/iceberg/src/arrow/reader.rs:
```diff
@@ -207,7 +233,8 @@ impl ArrowReader {
 
-        let delete_predicate = delete_file_manager.build_delete_predicate(task.schema.clone())?;
+        let delete_filter = delete_filter_rx.await.unwrap()?;
+        let delete_predicate = delete_filter.build_equality_delete_predicate(&task).await?;
```
Please can you explain a bit more? I don't understand what you're asking for here. As per the previous comment, we're using ParquetRecordBatchStream to do most of the heavy lifting for us on a number of features. This means that the design of the DeleteFilter can't be as simple as just exposing a method that filters a recordbatch stream.

sdd commented on crates/iceberg/src/arrow/delete_file_loader.rs:
```diff
@@ -0,0 +1,530 @@
+
+#[allow(unused)]
+#[derive(Clone, Debug)]
+pub(crate) struct CachingDeleteFileLoader {
```
I've extracted the base functionality out into `BasicDeleteFileLoader`. This is `pub`, and the original `CachingDeleteFileLoader` has stayed `pub(crate)` - `CachingDeleteFileLoader` is more tightly integrated with `ArrowReader` due to the usage of `ParquetRecordBatchStream`, whereas `BasicDeleteFileLoader` is aimed at external query engines.

sdd commented on crates/iceberg/src/arrow/delete_file_loader.rs:
```diff
@@ -0,0 +1,530 @@
+    InProgEqDel(EqDelFuture),
+    DelVecs(HashMap<String, DeleteVector>),
+    EqDel,
+}
```
Yes, they are still needed

sdd commented on crates/iceberg/src/arrow/delete_filter.rs:
```diff
@@ -0,0 +1,337 @@
+
+impl DeleteFilter {
+    /// Retrieve a delete vector for the data file associated with a given file scan task
+    pub fn get_delete_vector(
```
I've made all of `DeleteFilter` to be `pub(crate)` now then in that case. As per previous replies, `pub fn filter(input: ArrowRecordBatchStream)` does not make sense for usage with `ArrowReader`. If we also want to expose a public API for filtering that does implement that interface, we can add that in a follow-up PR.

sdd commented:
@liurenjie1024 back to you. I've addressed your suggestions that I think make sense to immediately change. I look forward to hearing back from you on the other points! :-)

dentiny commented on crates/iceberg/src/arrow/caching_delete_file_loader.rs:
```diff
@@ -0,0 +1,344 @@
+    ///      as per the other delete file types - only this time it is accompanied by a one-shot
+    ///      channel sender that we will eventually use to resolve the shared future that we stored
+    ///      in the state.
+    ///  * When this gets updated to add support for delete vectors, the load phase will return
```
Looking forward to the puffin / deletion vector support!

sdd replied:
Me too! 😁

xxchan commented on crates/iceberg/src/delete_file_index.rs:
```diff
@@ -199,18 +218,22 @@ pub(crate) struct DeletesForDataFile<'a> {
     state: Arc<RwLock<DeleteFileIndexState>>,
     data_file: &'a DataFile,
     seq_num: Option<i64>,
+    waker_set: Arc<WakerSet>,
 }
 
 impl Future for DeletesForDataFile<'_> {
     type Output = Result<Vec<FileScanTaskDeleteFile>>;
 
-    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         match self.state.try_read() {
             Ok(guard) => match guard.deref() {
                 DeleteFileIndexState::Populated(idx) => Poll::Ready(Ok(
                     idx.get_deletes_for_data_file(self.data_file, self.seq_num)
                 )),
-                _ => Poll::Pending,
+                _ => {
+                    self.waker_set.insert(cx);
+                    Poll::Pending
+                }
             },
             Err(err) => Poll::Ready(Err(Error::new(ErrorKind::Unexpected, err.to_string()))),
         }
```
As mentioned in #1323:
- manual `Future` implementation is error-prone. I think `async fn` should be preferred.
- Another bug here: the `try_read` may return a `TryLockError::WouldBlock`, which should not be an err. It's like a `Pending`.

sdd replied:
Hi @xxchan - thanks for the feedback. Good spot on the `WouldBlock` error, I've updated accordingly. Regarding the `Future` vs `async fn`, I agree that `async fn` has fewer sharp edges but conceptually in this case I think that `Future` is a better fit. Since the `Future` required in this particular case is not a particularly complex one I think it is justified. We're essentially waiting on something to be ready before we proceed with its value which is exactly the intended use case of a `Future`.

xxchan replied:
> We're essentially waiting on something to be ready before we proceed with its value which is exactly the intended use case of a Future.

async fn is just a Future, so I think this point doesn't justify the usage of a manual Future impl.

And actually for "waiting on something to be ready", I think it means we need a synchronization primitive. 
Here the async computation is not "owned" by the future, but we are just waiting for a notification.

sdd replied:
OK, fair point. Will refactor

sdd replied:
@xxchan How does it look now? I think you were right.

sdd commented:
Hi @liurenjie1024 / @Xuanwo / @xxchan.

This is now ready again for review after a refactor taking into account @xxchan's great feedback. I'll be on holiday for a week after today so it would be great if you guys could take a look. Thanks!
